// knex/index.js
const createStore = () => [];

class QueryBuilder {
  constructor(knexInstance, table) {
    this.knexInstance = knexInstance;
    this.table = table;
    this.condition = undefined;
  }

  where(condition) {
    this.condition = condition;
    return this;
  }

  first() {
    if (this.table === 'merchants' && this.condition?.phone) {
      return this.knexInstance.__store.find((record) => record.phone === this.condition.phone);
    }

    return undefined;
  }

  insert(payload) {
    if (this.table !== 'merchants') {
      return {
        returning: async () => [payload],
      };
    }

    const store = this.knexInstance.__store;
    const record = {
      id: `merchant-${store.length + 1}`,
      ...payload,
    };
    store.push(record);

    return {
      returning: async () => [record],
    };
  }
}

function knex() {
  const store = createStore();

  function query(table) {
    return new QueryBuilder(query, table);
  }

  query.__store = store;

  return query;
}

export default knex;
export { knex };
